optimade/server/entry_collections.py

```diff
@@ -93,40 +93,49 @@ def count(self, **kwargs):
         for k in list(kwargs.keys()):
             if k not in ("filter", "skip", "limit", "hint", "maxTimeMS"):
                 del kwargs[k]
+        if "filter" not in kwargs:  # "filter" is needed for count_documents()
+            kwargs["filter"] = {}
         return self.collection.count_documents(**kwargs)
 
     def find(
         self, params: Union[EntryListingQueryParams, SingleEntryQueryParams]
-    ) -> Tuple[List[EntryResource], bool, NonnegativeInt, set]:
+    ) -> Tuple[List[EntryResource], NonnegativeInt, bool, NonnegativeInt, set]:
         criteria = self._parse_params(params)
-        if isinstance(params, EntryListingQueryParams):
-            criteria_nolimit = criteria.copy()
-            del criteria_nolimit["limit"]
-            nresults_now = self.count(**criteria)
-            nresults_total = self.count(**criteria_nolimit)
-            more_data_available = nresults_now < nresults_total
-            data_available = nresults_total
-        else:
-            more_data_available = False
-            data_available = self.count(**criteria)
-            if data_available > 1:
-                raise HTTPException(
-                    status_code=404,
-                    detail=f"Instead of a single entry, {data_available} entries were found",
-                )
+
         all_fields = criteria.pop("fields")
         if getattr(params, "response_fields", False):
             fields = set(params.response_fields.split(","))
         else:
             fields = all_fields.copy()
+
         results = []
         for doc in self.collection.find(**criteria):
             results.append(self.resource_cls(**self.resource_mapper.map_back(doc)))
 
-        if isinstance(params, SingleEntryQueryParams):
+        if isinstance(params, EntryListingQueryParams):
+            nresults_now = len(results)
+            criteria_nolimit = criteria.copy()
+            criteria_nolimit.pop("limit", None)
+            data_returned = self.count(**criteria_nolimit)
+            more_data_available = nresults_now < data_returned
+        else:
+            # SingleEntryQueryParams, e.g., /structures/{entry_id}
+            data_returned = 1
+            more_data_available = False
+            if len(results) > 1:
+                raise HTTPException(
+                    status_code=404,
+                    detail=f"Instead of a single entry, {len(results)} entries were found",
+                )
             results = results[0] if results else None
 
-        return results, more_data_available, data_available, all_fields - fields
+        return (
+            results,
+            data_returned,
+            more_data_available,
+            self.count(),  # data_available
```

Comment:
```suggestion
            self.estimated_document_count(),  # data_available
```

Just checked the mongo docs and I think this is the new API call that optimises count based on metadata for our purposes (see the [pymongo docs](https://api.mongodb.com/python/current/api/pymongo/collection.html?highlight=document%20count#pymongo.collection.Collection.estimated_document_count)).

Reply:
This calls the local `count()` method, i.e., here I am calling `MongoCollection.count()`, which in turn calls the `mongomock` or `pymongo` collection's `count_documents()`. However, I can see we are using `estimated_document_count()` for the `__len__` built-in method.
So instead of returning extra variables, we can simply take the length of the collection as being `data_available`.

Reply:
Ah yeah, my bad, thought we were inheriting from the pymongo for some reason. Happy with this now!